src/components/SearchModal/CurrencySearch.tsx

```diff
@@ -100,34 +97,57 @@ export function CurrencySearch({
     )
   }, [chainId, data?.portfolios])
 
-  const sortedTokens: Token[] = useMemo(
-    () =>
-      !balancesAreLoading
-        ? filteredTokens
-            .filter((token) => {
-              if (onlyShowCurrenciesWithBalance) {
-                return balances[token.address?.toLowerCase()]?.usdValue > 0
-              }
+  const sortedTokens: Token[] = useMemo(() => {
+    const portfolioTokens = data?.portfolios?.[0].tokenBalances
+      ?.map((tokenBalance) => {
+        if (!tokenBalance?.token?.chain || !tokenBalance.token?.address || !tokenBalance.token?.decimals) {
+          return undefined
+        }
+        return new Token(
+          supportedChainIdFromGQLChain(tokenBalance.token?.chain) ?? ChainId.MAINNET,
+          tokenBalance.token?.address,
+          tokenBalance.token?.decimals,
+          tokenBalance.token?.symbol,
+          tokenBalance.token?.name
+        )
+      })
+      .filter((token) => !!token) as Token[]
+
+    const filteredTokens = Object.values(defaultTokens)
+      .filter(getTokenFilter(debouncedQuery))
+      // Filter out tokens with balances so they aren't duplicated when we merge below.
+      .filter((token) => !(token.address?.toLowerCase() in balances))
+    const mergedTokens = [...(portfolioTokens ?? []), ...filteredTokens]
+
+    if (balancesAreLoading) {
+      return mergedTokens
+    }
+
+    return mergedTokens
+      .filter((token) => {
+        if (onlyShowCurrenciesWithBalance) {
+          return balances[token.address?.toLowerCase()]?.usdValue > 0
+        }
+
+        // If there is no query, filter out unselected user-added tokens with no balance.
+        if (!debouncedQuery && token instanceof UserAddedToken) {
```

Comment:
I think that `UserAddedToken`s are only introduced through `defaultTokens`, so de-duping in L119 may cause them to continue to display with no balance.

Reply:
it is true that UserAddedTokens are introduced through defaultTokens, but i tested with a user-added token too and it's appearing correctly

Reply:
this is saying it should *not* appear if it has no balance. if that's still functional then lgtm.